### PR TITLE
record learning rate to debug log, change LearningRateSchedule to HyperParameterSchedule

### PR DESCRIPTION
**Pull request by qiuxin2012**
record learning rate to debug log for #257 

 change LearningRateSchedule to HyperParameterSchedule for #354

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 1]
```diff
@@ -69,24 +71,23 @@ class SGD[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T]
           add(ev.fromType[Double](1 - damp), dfdx)
       }
 
-      if (nesterov) {
+      if (cntv) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: cntv is a very confused name

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 2]
```diff
@@ -101,40 +102,64 @@ class SGD[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T]
...
+  case class EpochSchedule(regimes : Array[Regime]) extends HyperParameterScheduler {
     override def updateHyperParameter(config: Table, state: Table): Unit = {
       val epoch = config[Int]("epoch")
       for (r <- regimes) {
         if (epoch >= r.startEpoch && epoch <= r.endEpoch) {
           config.add(r.config)
         }
       }
-      config("clr") = -config.get[Double]("learningRate").getOrElse(1e-3)
+      config("clr") = config.get[Double]("learningRate").getOrElse(1e-3)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: why change this? If you want to display a positive learningRate, you should negative it at the display part.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 3]
```diff
@@ -101,40 +102,64 @@ class SGD[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T]
...
-  case class Poly(power : Double, maxIteration : Int) extends LearningRateSchedule {
+  case class Poly(power : Double, maxIteration : Int) extends HyperParameterScheduler {
     override def updateHyperParameter(config: Table, state: Table): Unit = {
       val lr = config.get[Double]("learningRate").getOrElse(1e-3)
       val nevals = state.get[Int]("evalCounter").getOrElse(0)
       val clr = if (nevals > maxIteration) {
         0.0
       } else {
-        -lr * math.pow(1.0 - nevals.toDouble / maxIteration, power)
+        lr * math.pow(1.0 - nevals.toDouble / maxIteration, power)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: why change this? If you want to display a positive learningRate, you should negative it at the display part.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 4]
```diff
@@ -101,40 +102,64 @@ class SGD[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T]
...
       config("clr") = clr
     }
   }
 
-  case class Step(stepSize : Int, gamma : Double) extends LearningRateSchedule {
+  case class Step(stepSize : Int, gamma : Double) extends HyperParameterScheduler {
     override def updateHyperParameter(config: Table, state: Table): Unit = {
       val lr = config.get[Double]("learningRate").getOrElse(1e-3)
-      var clr = -lr
+      var clr = lr
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: why change this? If you want to display a positive learningRate, you should negative it at the display part.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 5]
```diff
@@ -146,21 +171,21 @@ object SGD {
     }
   }
 
-  case class EpochDecay(decayType: (Int) => Double) extends LearningRateSchedule {
+  case class EpochDecay(decayType: (Int) => Double) extends HyperParameterScheduler {
     override def updateHyperParameter(config: Table, state: Table): Unit = {
       val lr = config.get[Double]("learningRate").getOrElse(1e-1)
-      var clr = -lr
+      var clr = lr
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: why change this? If you want to display a positive learningRate, you should negative it at the display part.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 6]
```diff
@@ -146,21 +171,21 @@ object SGD {
...
       config("clr") = clr
     }
   }
 
-  case class EpochStep(stepSize : Int, gamma : Double) extends LearningRateSchedule {
+  case class EpochStep(stepSize : Int, gamma : Double) extends HyperParameterScheduler {
     override def updateHyperParameter(config: Table, state: Table): Unit = {
       val lr = config.get[Double]("learningRate").getOrElse(1e-3)
-      var clr = -lr
+      var clr = lr
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: why change this? If you want to display a positive learningRate, you should negative it at the display part.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 7]
```diff
@@ -171,12 +196,12 @@ object SGD {
...
   }
 
-  case class Default() extends LearningRateSchedule {
+  case class Default() extends HyperParameterScheduler {
     override def updateHyperParameter(config: Table, state: Table): Unit = {
       val lr = config.get[Double]("learningRate").getOrElse(1e-3)
       val lrd = config.get[Double]("learningRateDecay").getOrElse(0.0)
       val nevals = state.get[Int]("evalCounter").getOrElse(0)
-      config("clr") = -lr / (1 + nevals * lrd)
+      config("clr") = lr / (1 + nevals * lrd)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: why change this? If you want to display a positive learningRate, you should negative it at the display part.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 8]
```diff
@@ -246,6 +247,29 @@ object DistriOptimizer {
         logger.debug("Dropped modules: " + (driverSubModelNum - finishedModelNum))
         lossArray = new Array[Double](_subModelNumber)
 
+        // record the hyper parameters to debug log, only support SGD now.
+        optimMethod match {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: please move these into a private method

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 9]
```diff
@@ -246,6 +247,29 @@ object DistriOptimizer {
...
+          case sgd: SGD[T] =>
+            val hpScheduler = state.getOrElse[HyperParameterScheduler](
+              "hyperParameterScheduler", Default())
+            hpScheduler.updateHyperParameter(driverState, state)
+            logger.debug(s"Current learning rate is ${driverState.getOrElse[Double]("clr",
+              driverState.getOrElse[Double]("learningRate", 1e-3))}")
+            if (driverState.contains("wd")) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: First, there may be only weightDecays, so no wd. Do we ignore this?

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 10]
```diff
@@ -246,6 +247,29 @@ object DistriOptimizer {
...
+              logger.debug(s"Current weight decay is ${driverState[Double]("wd")}")
+            }
+            if (driverState.contains("mom")) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: same with above

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 11]
```diff
@@ -246,6 +247,29 @@ object DistriOptimizer {
...
+              logger.debug(s"Current momentum is ${driverState[Double]("mom")}")
+            }
+            if (driverState.contains("damp")) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: same with above

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 12]
```diff
@@ -246,6 +247,29 @@ object DistriOptimizer {
...
+              logger.debug(s"Current damping is ${driverState[Double]("damp")}")
+            }
+            if (driverState.contains("cntv")) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: same with above

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 13]
```diff
@@ -32,17 +32,19 @@ class SGD[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T]
...
-    val mom = config.get[Double]("momentum").getOrElse(0.0)
-    val damp = config.get[Double]("dampening").getOrElse(mom)
-    val nesterov = config.get[Boolean]("nesterov").getOrElse(false)
-    val lrs = config.get[Tensor[T]]("learningRates").getOrElse(null)
-    val wds = config.get[Tensor[T]]("weightDecays").getOrElse(null)
-
-    require(!nesterov || (mom > 0 && damp == 0),
+    val scheduler = config.getOrElse[HyperParameterScheduler]("hyperParameterScheduler", Default())
+    scheduler.updateHyperParameter(config, _state)
+
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: We should not give two interface(e.g. wd and weightDecay for weightDecay), or the code is confused and easy to make bug.

If you touch other hyperparameter instead of only learningRate, you should make the interface align with learningRate, like a clr and learningRate, cweightDecay and weightDecay, etc.

So SGD only read hyperparameter from clr, cweightDecay... And user specify learningRate and weightDecay in their configuration

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 9 above)
Comment (qiuxin2012) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: no wd means no change, so we can ignore this.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 2 above)
Comment (qiuxin2012) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: clr means "current learning rate", right? Learning rate is a positive value, and why clr is a negative value? It's confused when user write user defined HyperParamterScheduler like Poly2 or Step2.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 13 above)
Comment (qiuxin2012) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: If we only read hyperparameter from clr, cweightDecay..., then when the user are writing their user-defined HyperParameterScheduler, he needs to write more codes to define cweightDecay, clearningRateDecay... even he doesn't change the weightDecay and learningRateDecay.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 14]
```diff
@@ -246,6 +247,29 @@ object DistriOptimizer {
         logger.debug("Dropped modules: " + (driverSubModelNum - finishedModelNum))
         lossArray = new Array[Double](_subModelNumber)
 
+        // record the hyper parameters to debug log, only support SGD now.
+        optimMethod match {
+          case sgd: SGD[T] =>
+            val hpScheduler = state.getOrElse[HyperParameterScheduler](
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: why not just print state? If there's new member in state, you have to change this piece of code.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 2 above)
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Here all other hp are missed

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 13 above)
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: don't mix them together.
